compiler/main/PhaseTracker.h

```diff
@@ -22,153 +22,183 @@
+ * a) Create a Timer object that will record the wall time as early in the     *
+ *    compiler as possible                                                     *
+ *                                                                             *
+ * b) Use the Timer to determine the elapsed time between the start of the     *
+ *    program and the start of each phase                                      *
+ *                                                                             *
+ * c) Compute the elapsed time for each phase                                  *
+ *                                                                             *
+ * d) Stop the timer and generate one or more reports on the time spent in     *
+ *    each phase.
+ *                                                                             *
+ * Every phase has a descriptive name that will be used by the reports.        *
+ *                                                                             *
+ * The main loop for the compiler, in runpasses.cpp, has a notion of a "pass"  *
+ * which consists of three phases;                                             *
+ *                                                                             *
+ *    1) The primary computation                                               *
+ *    2) An optional verify / check phase                                      *
+ *    3) An AST cleanup phase                                                  *
+ *                                                                             *
+ * The tracker collects the sub-phases and reports the overall time in terms   *
+ * of these passes.  Phases that occur before and after the Passes ignore      *
+ * the check and clean phases.                                                 *
+ *                                                                             *
+ ************************************** | *************************************/
 
 class Phase;
 class Pass;
 
-class PhaseTracker
-{
+struct PhaseData {
+  unsigned long                timeInUsecs;
+  MemoryTracker::MemoryInBytes memory;
+
+  PhaseData() : timeInUsecs(0), memory(0) {}
+  PhaseData(unsigned long time, MemoryTracker::MemoryInBytes memory)
+      : timeInUsecs(time), memory(memory) {}
+
+  std::string serialize() const {
+    char buffer[256];
+    snprintf(buffer, sizeof(buffer), "%lu %lld", timeInUsecs, memory);
+    return std::string(buffer);
+  }
+  static PhaseData deserialize(std::string_view str) {
+    PhaseData result;
+    sscanf(str.data(), "%lu %lld", &result.timeInUsecs, &result.memory);
+    return result;
+  }
+
+  PhaseData operator-(const PhaseData& other) const {
+    return PhaseData(this->timeInUsecs - other.timeInUsecs,
+                     this->memory - other.memory);
+  }
+  PhaseData operator+(const PhaseData& other) const {
+    return PhaseData(this->timeInUsecs + other.timeInUsecs,
+                     this->memory + other.memory);
```

Comment:
I think you can define this as `auto ret = *this; return ret += other` to avoid duplication.